presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PlanNodeDecorrelator.java

```diff
@@ -189,6 +218,78 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
                     true));
         }
 
+        private Optional<DecorrelationResult> rewriteLimitWithRowCountGreaterThanOne(DecorrelationResult childDecorrelationResult, long limit, PlanNodeId nodeId)
+        {
+            PlanNode decorrelatedChildNode = childDecorrelationResult.node;
+            Set<Symbol> constantSymbols = childDecorrelationResult.getConstantSymbols();
+
+            if (constantSymbols.isEmpty() || !constantSymbols.containsAll(childDecorrelationResult.symbolsToPropagate)) {
+                return Optional.empty();
+            }
+
+            // rewrite Limit to RowNumberNode partitioned by constant symbols
+            RowNumberNode rowNumberNode = new RowNumberNode(
+                    nodeId,
+                    decorrelatedChildNode,
+                    ImmutableList.copyOf(childDecorrelationResult.symbolsToPropagate),
+                    symbolAllocator.newSymbol("row_number", BigintType.BIGINT),
+                    Optional.of(toIntExact(limit)),
+                    Optional.empty());
+
+            return Optional.of(new DecorrelationResult(
+                    rowNumberNode,
+                    childDecorrelationResult.symbolsToPropagate,
+                    childDecorrelationResult.correlatedPredicates,
+                    childDecorrelationResult.correlatedSymbolsMapping,
+                    false));
+        }
+
+        @Override
+        public Optional<DecorrelationResult> visitTopN(TopNNode node, Void context)
+        {
+            if (node.getCount() == 0) {
```

Comment:
`LIMIT 0` is legal. Anyway it should be trimmed before decorrelation.

Reply:
maybe you could use `checkState(node.getCount() !=  0, "LIMIT 0 should be trimmed before decorrelation")`?

Reply:
> maybe you could use checkState(node.getCount() != 0, "LIMIT 0 should be trimmed before decorrelation")?

Since there's no guarantee rules are applied in any particular order, that would force the caller to do this check and skip calling this method, so not sure we gain anything. Since we already use `Optional.empty()` to signal "couldn't do anything", I'd keep it that way.